contributors/design-proposals/daemonset-update.md

```diff
@@ -165,22 +187,38 @@ DaemonSets on the apiserver.
 
 For each pending DaemonSet updates, it will:
 
-1. Find all pods whose labels are matched by DaemonSet `.spec.selector`. 
-   - If `OwnerReference` is implemented for DaemonSets, filter out pods that
-     aren't controlled by this DaemonSet too.
+1. Find all pods controlled by DaemonSet
 1. Check `DaemonSetUpdateStrategy`:
    - If `OnDelete`: do nothing
    - If `RollingUpdate`:
-     - Find all daemon pods that belong to this DaemonSet using label selectors.
-       - Pods with "pod-template-generation" value equal to the DaemonSet's
-         `.spec.templateGeneration` are new pods, otherwise they're old pods.
-       - Note that pods without "pod-template-generation" labels (e.g. DaemonSet
-         pods created before RollingUpdate strategy is implemented) will be
-         seen as old pods.
+     - Find existing PodTemplates owned by DaemonSet, and sort them by the value
+       of `pod-template-generation` label
+     - Clean up PodTemplates based on `.spec.revisionHistoryLimit`
+       - Always keep the PodTemplate if its `pod-template-generation` label has
+         the same value of any existing pods owned by the DaemonSet
+       - Remove PodTemplates from the ones with the smallest `pod-template-generation`
+         to the highest, until `.spec.revisionHistoryLimit` is hit or no
+         PodTemplates are allowed to be removed 
+     - Create a PodTemplate based on DaemonSet
```

Comment:
I think it's highly unlike that this update pattern will run more than once for the same template but we can solve that case by not relabeling in case a previous pod-template-generation label already exists in the PodTemplate. Assuming your example, all Pods would be labeled as pod-template-generation=1, even though ds.spec.generation=5. The DS template matches with the 1st PodTemplate so we simply continue to mint Pods from that template w/o any changes to it and its existing Pods. Reminder: the original reason for using templateGeneration was to avoid naming collisions introduced by hashing.